docker/flink/Dockerfile

```diff
@@ -1,14 +1,141 @@
-ARG FLINK_VERSION=1.18.1
+ARG FLINK_VERSION=1.19.0
 FROM flink:${FLINK_VERSION}
 
 # ibis-flink requires PyFlink dependency
-ARG FLINK_VERSION=1.18.1
+ARG FLINK_VERSION=1.19.0
 RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/flink/flink-python/${FLINK_VERSION}/flink-python-${FLINK_VERSION}.jar
 
-# install python3 and pip3
-RUN apt-get update -y && \
-    apt-get install -y python3 python3-pip python3-dev openjdk-11-jdk-headless && \
-    rm -rf /var/lib/apt/lists/*
+# install python3 and pip3, following https://github.com/docker-library/python/blob/1b7a110/3.11/bookworm/Dockerfile
+# add build-essential, libbz2-dev, libffi-dev, liblzma-dev, libssl-dev, and openjdk-11-jdk-headless
+
+# ensure local python is preferred over distribution python
+ENV PATH /usr/local/bin:$PATH
+
+# cannot remove LANG even though https://bugs.python.org/issue19846 is fixed
+# last attempted removal of LANG broke many users:
+# https://github.com/docker-library/python/pull/570
+ENV LANG C.UTF-8
+
+# runtime dependencies
+RUN set -eux; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends \
+        build-essential \
+		libbluetooth-dev \
+        libbz2-dev \
+        libffi-dev \
+        liblzma-dev \
+        libssl-dev \
+        openjdk-11-jdk-headless \
+		tk-dev \
+		uuid-dev \
+	; \
+	rm -rf /var/lib/apt/lists/*
+
+# set GPG_KEY and PYTHON_VERSION corresponding to the desired Python version
+# https://github.com/docker-library/python/blob/0d53911/Dockerfile-linux.template#L54-L81
+ARG GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D
+ARG PYTHON_VERSION=3.11.8
+
+RUN set -eux; \
+	\
+	wget -O python.tar.xz "https://www.python.org/ftp/python/${PYTHON_VERSION%%[a-z]*}/Python-$PYTHON_VERSION.tar.xz"; \
+	wget -O python.tar.xz.asc "https://www.python.org/ftp/python/${PYTHON_VERSION%%[a-z]*}/Python-$PYTHON_VERSION.tar.xz.asc"; \
+	GNUPGHOME="$(mktemp -d)"; export GNUPGHOME; \
+	gpg --batch --keyserver hkps://keys.openpgp.org --recv-keys "$GPG_KEY"; \
+	gpg --batch --verify python.tar.xz.asc python.tar.xz; \
+	gpgconf --kill all; \
+	rm -rf "$GNUPGHOME" python.tar.xz.asc; \
+	mkdir -p /usr/src/python; \
+	tar --extract --directory /usr/src/python --strip-components=1 --file python.tar.xz; \
+	rm python.tar.xz; \
+	\
+	cd /usr/src/python; \
+	gnuArch="$(dpkg-architecture --query DEB_BUILD_GNU_TYPE)"; \
+	./configure \
+		--build="$gnuArch" \
+		--enable-loadable-sqlite-extensions \
+		--enable-optimizations \
+		--enable-option-checking=fatal \
+		--enable-shared \
+		$( (( $PYTHON_VERSION == '3.11.8' )) && printf %s '--with-lto' ) \
+		--with-system-expat \
+		--without-ensurepip \
+	; \
+	nproc="$(nproc)"; \
+	EXTRA_CFLAGS="$(dpkg-buildflags --get CFLAGS)"; \
+	LDFLAGS="$(dpkg-buildflags --get LDFLAGS)"; \
+	make -j "$nproc" \
```

Comment:
Why do we need to build Python from scratch?

Reply:
I think the issue is that it needs to be the default Python on the remote image, coming back to

> BTW I think this could potentially be a lot easier if PyFlink supported a PYFLINK_PYTHON env var with path to executable, like PySpark does; can't verify, though.

Reply:
default in this case would mean that it's the python executable that has the highest precedence in the `PATH`, so I think activating a micromamba environment will do the trick.  Unless Flink is looking for a hardcoded path, in which case, they should be punished.

Reply:
> default in this case would mean that it's the python executable that has the highest precedence in the `PATH`, so I think activating a micromamba environment will do the trick. Unless Flink is looking for a hardcoded path, in which case, they should be punished.

Maybe you're right? I found this: https://github.com/apache/flink/blob/3841f0622552674fe0013b89605594a401cb6902/flink-python/src/main/java/org/apache/flink/python/PythonOptions.java#L159

I don't completely understand/recall why it was failing when I installed with deadsnakes, though, and linked the installed Python to `/usr/bin/python`. Alternatively, maybe this option means can pass the interpreter...

Reply:
Actually, whatever I have now seems to pass; let me see if it's acceptable first, before sinking more time into this. :)